text/0000-natvis.md

```diff
@@ -0,0 +1,428 @@
+    let rx = Regex::new(r#"^#define\s+(\w+)\s+([0-9]+)\s*(//(.*))?"#).unwrap();
+    for captures in rx.captures_iter(input) {
+        let my_match: Match = captures.get(1).unwrap();
+        do_some_work(my_match.as_str());
+    }
+}
+```
+
+Let's say that Carol is debugging the app, there's a problem in
+`do_some_work()`. (Perhaps some code path has triggered a panic.) Carol wants
+to look at the state of the app, inside the `find_c_defines` function,
+and she specifically wants to see what the state of `captures` is. So she
+selects the `find_c_defines` call frame and looks at the local variables
+window.
+
+Unfortunately, the debugger's view of the `captures` variable does not give
+her any useful information at all. It shows only something like:
+
+```text
+> Variables
+  > captures: {...}
+    > text: "...the entire input text..."
+    > locs: {...}
+      > __0: (4) vec![None, None, None, None]
+      > named_groups: (refs:2) size=0, capacity=1
+        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
+          > ptr: {pointer:0xNNNNNNNN}
+            > pointer: {...}
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is useless.
+Even the implementor of `regex` would have a hard time knowing how to decode
+this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > $1: "SOME_CONSTANT"
+    > $2: "42"
+    > $3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis visualizations which is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger.
+
+It should be easy for crate developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis XML files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+When writing Natvis files for C++, developers write a standalone XML document
+and add it to their project. The build system (such as Visual Studio) knows
```

Comment:
"add it to their project" what does "add" mean here? What is the "project"?

Reply:
The phrasing is fairly standard in the C++ / Visual Studio community.

Reply:
Sure, but this is not the C++ / Visual Studio community, so it would be good to explain it :-)

Reply:
To clarify, I'm trying to understand exactly how natvis files are handled by VS - does this mean that VS stores a record of the filename in some project metadata (roughly equivalent to a manifest file in that it would be committed to version control, etc) and that this record is used when building the project? Although this sounds kind of obvious, one could imagine alternatives such as the project being a developer's local view, or "adding" a natvis file meaning that it is copied into source code or compiled into an intermediate format or something.

Reply:
> "add it to their project" what does "add" mean here? What is the "project"?

I've updated this information and described what it would mean to add Natvis to a project within Visual Studio and refactored the structure of this.